### PR TITLE
Thumbnails

### PR DESCRIPTION
**Pull request by zachmullen**

### CODE REVIEW

**Comment by zachmullen** [comment]
This massive and horrible branch is finally ready for review. In addition to adding the new features of local jobs and a thumbnailing plugin, it also addresses several core issues and problems in our testing infrastructure.

**Comment by cpatrick** [comment]
It would be great to think of good was of extending this to support more exotic thumbnail types. From ITK images, VTK for geometry, etc.

**Comment by zachmullen** [comment]
@cpatrick That last commit addresses your comment.

**Comment by cpatrick** [comment]
This seems to break with the provenance plugin:

```
message: "Exception: Exception('Attempt to load null ObjectId: None',)"
trace: [["girder/api/rest.py", 278, "endpointDecorator", "val = fun(self, args, kwargs)"],…]
0: ["girder/api/rest.py", 278, "endpointDecorator", "val = fun(self, args, kwargs)"]
1: ["girder/api/rest.py", 772, "POST", "return self.handleRoute(method, path, params)"]
2: ["girder/api/rest.py", 560, "handleRoute", "val = handler(**kwargs)"]
3: ["girder/api/access.py", 50, "accessDecorator", "return fun(*args, **kwargs)"]
4: ["girder/api/rest.py", 226, "wrapped", "return fun(*args, **kwargs)"]
5: ["/Users/cpatrick/Projects/Kitware/General/girder/plugins/thumbnails/server/rest.py", 71,…]
6: ["/Users/cpatrick/Projects/Kitware/General/girder/plugins/jobs/server/models/job.py", 214,…]
7: ["girder/events.py", 241, "trigger", "handler['handler'](e)"]
8: ["/Users/cpatrick/Projects/Kitware/General/girder/plugins/jobs/server/__init__.py", 43,…]
9: ["/Users/cpatrick/Projects/Kitware/General/girder/plugins/thumbnails/server/worker.py", 36, "run",…]
10: ["/Users/cpatrick/Projects/Kitware/General/girder/plugins/thumbnails/server/worker.py", 118,…]
11: ["girder/models/upload.py", 64, "handleChunk", "return self.finalizeUpload(upload, assetstore)"]
12: ["girder/models/upload.py", 122, "finalizeUpload", "self.model('file').save(file)"]
13: ["girder/models/model_base.py", 306, "save", "document)"]
14: ["girder/events.py", 241, "trigger", "handler['handler'](e)"]
15: ["/Users/cpatrick/Projects/Kitware/General/girder/plugins/provenance/server/resource.py", 364,…]
16: ["girder/models/item.py", 124, "load", "exc=exc)"]
17: ["girder/models/model_base.py", 390, "load",…]
```

**Comment by cpatrick** [comment]
image

We may want a better error message in the presence of non-numeric input.

File: clients/web/src/views/widgets/SearchFieldWidget.js [hunk 1]
```diff
@@ -55,7 +56,8 @@ girder.views.SearchFieldWidget = girder.View.extend({
                     this.trigger('g:resultClicked', {
                         type: link.attr('resourcetype'),
                         id: link.attr('resourceid'),
-                        text: link.text()
+                        text: link.text().trim(),
+                        icon: link.attr('g-icon')
                     });
```
Comment (cpatrick) [inline clients/web/src/views/widgets/SearchFieldWidget.js]: Make a utility function  for this :meat_on_bone:

File: girder/models/model_base.py [hunk 2]
```diff
@@ -180,7 +186,7 @@ def validate(self, doc):
         :type doc: dict
         """
         raise Exception('Must override validate() in %s model.'
-                        % self.__class__.__name__)  # pragma: no cover
+                        % self.__name__)  # pragma: no cover
```
Comment (cpatrick) [inline girder/models/model_base.py]: This is a bug.

File: girder/models/model_base.py [hunk 3]
```diff
@@ -188,7 +194,7 @@ def initialize(self):
         self.name. Also, they should set any indexed fields that they require.
         """
         raise Exception('Must override initialize() in %s model'
-                        % self.__class__.__name__)  # pragma: no cover
+                        % self.__name__)  # pragma: no cover
```
Comment (cpatrick) [inline girder/models/model_base.py]: This is a bug.

**Comment by cpatrick** [comment]
Would be nice from a UI perspective to have a Thumbnail header with icon a la metadata.

**Comment by cpatrick** [comment]
Review pass complete.

**Comment by zachmullen** [comment]
ping @cpatrick I think that addresses all of the QA findings

**Comment by zachmullen** [comment]
@cpatrick any other comments on this branch or is it ready to merge?

**Comment by cpatrick** [comment]
[LGTM](http://lgtm.in/i/QgJeRiGwn)

**Pull request merged**